### PR TITLE
scripts: small refactorings

### PR DESCRIPTION
**Pull request by georglauterbach**
# Description

The commit descriptions are _somewhat_ detailed. It basically boils down to:

1. Cron scripts are now added with HEREDOCs
2. Minor style adjustments

## Type of change

- [x] Improvement (non-breaking change that does improve existing functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [x] If necessary I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by polarathene — CHANGES_REQUESTED** [review]
Good PR overall! :+1: 

- Some minor change suggestions if you think any would be improvements.
- Caution towards the SSL `check-for-changes.sh` change, may regress support for other users without third-party customizations.
- `log.sh` should have explicit `error` case, `always` is better suited to use the `*` fallback. Not sure why `error` is set as default `LOG_LEVEL` case to rely on `*` as fallback/highlighting?
- Possible bug with one of your cron heredocs.

### Feedback: Personal opinion - Not a fan of HEREDOC

Cron jobs to heredocs ([relevant commit with commit message for context](https://github.com/docker-mailserver/docker-mailserver/pull/2485/commits/172e50dcc175a301fa10b4ad6c2a382b902bff4d)) is not something I'm fond of personally.

I prefer external files instead of embedded/inlined, ENV as a convenience isn't sufficient for me. We already write ENV to a file for other scripts to read and utilize. I'd rather cron jobs we add are all in a common location from a maintenance perspective, but I understand if the majority of maintainers prefer the heredoc approach instead and have no interest in fighting that decision.

---

### Feedback: Log changes could have been covered in a separate PR

The log level change is rather large and seems it would have suited it's own PR rather than mixed in with the other changes covered ([relevant commit covering the bulk of logging changes](https://github.com/docker-mailserver/docker-mailserver/pull/2485/commits/abf18e534383cffa2e52f9a635b973a5f4a8751e)). I understand it's mostly isolated to the linked commit, but for review purposes separate PRs would have been nicer :sweat_smile:

File: .github/ISSUE_TEMPLATE/bug_report.yml
Comment (georglauterbach) [reply .github/ISSUE_TEMPLATE/bug_report.yml:135]: Done in a13e75d7924c2e4a653736f374eaeffcf8d2d254 on branch `update/log` - can be found in #2486.

File: docs/content/contributing/issues-and-pull-requests.md
Comment (georglauterbach) [reply docs/content/contributing/issues-and-pull-requests.md]: > Any reason for dropping the increased verbosity log advice instead of adapting it to the logging changes?

No particular reason, but the issue template and the docs already make it clear.

File: mailserver.env
Comment (georglauterbach) [reply mailserver.env]: Done in a13e75d7924c2e4a653736f374eaeffcf8d2d254 on branch `update/log` - can be found in #2486.

File: target/scripts/check-for-changes.sh
Comment (georglauterbach) [reply target/scripts/check-for-changes.sh]: Done in a13e75d7924c2e4a653736f374eaeffcf8d2d254 on branch `update/log` - can be found in #2486.

File: target/scripts/check-for-changes.sh
Comment (georglauterbach) [reply target/scripts/check-for-changes.sh]: You are, indeed, very correct. I'm not sure why I missed this (especially as I implemented it myself). This is now reverted and will stay this way. Thank you for noticing, I'm currently not on point with this...

File: test/mail_special_use_folders.bats
Comment (georglauterbach) [reply test/mail_special_use_folders.bats]: Should be corrected now.

File: target/scripts/start-mailserver.sh
Comment (georglauterbach) [reply target/scripts/start-mailserver.sh]: Done in a13e75d7924c2e4a653736f374eaeffcf8d2d254 on branch `update/log` - can be found in #2486.

File: target/scripts/startup/setup-stack.sh
Comment (georglauterbach) [reply target/scripts/startup/setup-stack.sh:1118]: Good catch!

File: target/scripts/startup/setup-stack.sh
Comment (georglauterbach) [reply target/scripts/startup/setup-stack.sh]: Done in 6dc28be296d45851680b0751a25b9309688651dd

File: target/scripts/startup/setup-stack.sh
Comment (georglauterbach) [reply target/scripts/startup/setup-stack.sh]: Done in 6dc28be296d45851680b0751a25b9309688651dd

**Comment by georglauterbach** [comment]
> Good PR overall! +1

Thanks!

---

> * Some minor change suggestions if you think any would be improvements.
> * [Caution towards the SSL `check-for-changes.sh` change](https://github.com/docker-mailserver/docker-mailserver/pull/2485#discussion_r826486054). It may regress support for other users without third-party customizations.
> * [`log.sh` should have explicit `error` case](https://github.com/docker-mailserver/docker-mailserver/pull/2485#discussion_r826457218), `always` is better suited to use the `*` fallback. Not sure why `error` is set as default `LOG_LEVEL` case to rely on `*` as fallback/highlighting?
> * [Possible bug with one of your cron heredoc](https://github.com/docker-mailserver/docker-mailserver/pull/2485#discussion_r826472255) additions.

All resolved :D

---
 
> ### Feedback: Personal opinion - Not a fan of HEREDOC
> 
> Cron jobs to heredocs ([relevant commit with commit message for context](https://github.com/docker-mailserver/docker-mailserver/pull/2485/commits/172e50dcc175a301fa10b4ad6c2a382b902bff4d)) is not something I'm fond of personally.
> 
> I prefer external files instead of embedded/inlined, ENV as a convenience isn't sufficient for me. We already write ENV to a file for other scripts to read and utilize. I'd rather cron jobs we add are all in a common location from a maintenance perspective, but I understand if the majority of maintainers prefer the heredoc approach instead and have no interest in fighting that decision.

I have to admit I disagree, obviously (otherwise I wouldn't have done it). I first tried your approach to see how it goes, but then I noticed that it is way more work to handle these scripts if they are not "inlined" into `setup-stack.sh` (with a HEREDOC). Moreover, we only actually have two cron scripts that are added. From experience with this project and writing many many lines of code, I would **strongly** advise against dedicated scripts. 

---

> ### Feedback: Log changes could have been covered in a separate PR

I agree. Therefore, I opened #2486.

File: target/scripts/helpers/error.sh
Comment (casperklein) [inline target/scripts/helpers/error.sh:5]: Analog to `setup.sh` and other places, we should use variables for colors, instead of cryptic escape codes.

What about putting this:

https://github.com/docker-mailserver/docker-mailserver/blob/b730942b96aec3ab52699b5d927577f42eef0a45/setup.sh#L22-L26

globally in, e.g. `log.sh`?

File: target/scripts/helpers/error.sh
Comment (casperklein) [inline target/scripts/helpers/error.sh:75]: `SIGTERM` is the default. I am a fan of just providing what's necessary as argument.

File: target/scripts/helpers/error.sh
Comment (georglauterbach) [reply target/scripts/helpers/error.sh:5]: Probably a good idea! I will tend to it.

File: target/scripts/helpers/error.sh
Comment (casperklein) [reply target/scripts/helpers/error.sh:72]: How do you test this? This definitely works. Think about starting the container without a mail account. DMS keeps restarting in a loop --> stopped by `_shutdown` and then restarted via docker-compose policy `restart: always`.

Comment (georglauterbach) [reply target/scripts/helpers/error.sh:72]: I started DMS with `docker run -it --rm mailserver/docker-mailserver` and the container does not stop. PID 1 is still alive... I was curious about this too, as I too thought `kill` would by default send `SIGTERM`.

Comment (casperklein) [reply target/scripts/helpers/error.sh:72]: Does nothing happens or what is the result? What does `docker logs` shows, after executing `kill 1`.

`docker exec -it mailserver kill 1` works for me, as expected.

Comment (georglauterbach) [reply target/scripts/helpers/error.sh:72]: I thought DMS would exit automatically if there are no accounts, but the container just does not stop. Maybe there is a misconception on my side.

File: target/scripts/helpers/error.sh
Comment (polarathene) [inline target/scripts/helpers/error.sh:74]: Residue? It seems like the changes in full were carried over to https://github.com/docker-mailserver/docker-mailserver/pull/2486 and no longer belongs here? (although you dropped the `kill 1` in the other PR)

File: .github/ISSUE_TEMPLATE/bug_report.yml
Comment (polarathene) [inline .github/ISSUE_TEMPLATE/bug_report.yml:156]: Looks like this was all migrated to the other PR and doesn't belong here? (_I'm not sure if the line range is showing up for this comment, but all changes in this file_)

**Review by polarathene — CHANGES_REQUESTED** [review]

**Comment by casperklein** [comment]
> Feedback: Personal opinion - Not a fan of HEREDOC

Depends on the use-case IMO. Regarding the usage for our cron jobs, I think this is a good solution.

File: target/scripts/helpers/error.sh
Comment (georglauterbach) [reply target/scripts/helpers/error.sh:5]: Yes, I just did not get around to it yesterday (was interrupted in the process). I will take care of it.

File: .github/ISSUE_TEMPLATE/bug_report.yml
Comment (georglauterbach) [reply .github/ISSUE_TEMPLATE/bug_report.yml:156]: - There are more leftovers than I thought, this s also one...

File: target/scripts/helpers/error.sh
Comment (georglauterbach) [reply target/scripts/helpers/error.sh]: Done in 3c7483149d53e750925fe5651df3fadac1ae239e

File: .github/ISSUE_TEMPLATE/bug_report.yml
Comment (georglauterbach) [reply .github/ISSUE_TEMPLATE/bug_report.yml]: Done in 3c7483149d53e750925fe5651df3fadac1ae239e

File: target/scripts/helpers/error.sh
Comment (georglauterbach) [reply target/scripts/helpers/error.sh]: Done in 3c7483149d53e750925fe5651df3fadac1ae239e

File: target/scripts/startup/setup-stack.sh
Comment (polarathene) [inline target/scripts/startup/setup-stack.sh:1118]: I assume you want consistency with case syntax here?

```suggestion
    ( 'none' )
```

File: test/mail_ssl_manual.bats
Comment (polarathene) [inline test/mail_ssl_manual.bats:104]: Is this change relevant to this PR or the logs?

File: target/scripts/startup/setup-stack.sh
Comment (polarathene) [inline target/scripts/startup/setup-stack.sh:26]: I'm not able to highlight the line below with the `* )` case, but I assume you want that to be consistent and have a `(` prefix too?

**Review by polarathene — APPROVED** [review]
Possibly some changes you still want to make, otherwise LGTM :+1:

File: test/mail_ssl_manual.bats
Comment (georglauterbach) [reply test/mail_ssl_manual.bats:104]: Rather this PR :)

File: target/scripts/startup/setup-stack.sh
Comment (georglauterbach) [reply target/scripts/startup/setup-stack.sh:1118]: Yep, I want this to be in this PR as well :D I will commit this hand with the suggestion below.

Comment (georglauterbach) [reply target/scripts/startup/setup-stack.sh]: Done in d2c9aeda3725e6963105d845520d3a2e07a7aeac

File: target/scripts/startup/setup-stack.sh
Comment (georglauterbach) [reply target/scripts/startup/setup-stack.sh:26]: Done in d2c9aeda3725e6963105d845520d3a2e07a7aeac

**Comment by georglauterbach** [comment]
Thanks for the good feedback, very much appreciated!

**Review by polarathene — DISMISSED** [review]

**Review by polarathene — APPROVED** [review]

**Review by casperklein — APPROVED** [review]

**Pull request merged**
